algorithms/linfa-clustering/src/optics/algorithm.rs

```diff
@@ -0,0 +1,291 @@
+            core_distance: self.c_distance.map(|x| x.0),
+        }
+    }
+}
+
+impl<'a, F: Float> Eq for Neighbor<'a, F> {}
+
+impl<'a, F: Float> PartialEq for Neighbor<'a, F> {
+    fn eq(&self, other: &Self) -> bool {
+        self.r_distance == other.r_distance
+    }
+
+    fn ne(&self, other: &Self) -> bool {
+        !self.eq(other)
+    }
+}
+
+impl<'a, F: Float> PartialOrd for Neighbor<'a, F> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        self.r_distance.partial_cmp(&other.r_distance)
+    }
+}
+
+impl<'a, F: Float> Ord for Neighbor<'a, F> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.r_distance.cmp(&other.r_distance)
+    }
+}
+
+impl Optics {
+    pub fn params(min_points: usize) -> OpticsHyperParamsBuilder {
+        OpticsHyperParams::new(min_points)
+    }
+}
+
+impl<'a, F: Float, D: Data<Elem = F>> Transformer<&'a ArrayBase<D, Ix2>, OpticsAnalysis<'a, F>>
+    for OpticsHyperParams
+{
+    fn transform(&self, observations: &'a ArrayBase<D, Ix2>) -> OpticsAnalysis<'a, F> {
+        let mut result = OpticsAnalysis { orderings: vec![] };
+
+        let mut points = observations
+            .axis_iter(Axis(0))
+            .enumerate()
+            .map(|(i, x)| Neighbor::new(i, x))
+            .collect::<Vec<_>>();
+
+        // The BTreeSet is used so that the indexes are ordered to make it easy to find next
+        // index
+        let mut processed = BTreeSet::new();
+        let mut index = 0;
+        loop {
+            if index == points.len() {
+                break;
+            } else if processed.contains(&index) {
+                index += 1;
+                continue;
+            }
+            let mut expected = index + 1;
+            let mut points_index = 0;
+            for index in processed.range((index + 1)..) {
+                if expected != *index {
+                    points_index = expected;
+                    processed.insert(expected);
+                    break;
+                }
+                expected += 1;
+            }
+            index += 1;
+            let neighbors = find_neighbors(
+                &points[points_index].observation,
+                observations,
+                self.tolerance(),
+            );
+            let n = &mut points[points_index];
+            n.set_core_distance(self.minimum_points(), &neighbors);
+            result.orderings.push(n.sample());
+            if n.c_distance.is_some() {
+                // Here we get a list of "density reachable" samples that haven't been processed
+                // and sort them by reachability so we can process the closest ones first.
+                let mut seeds = Vec::new();
+                get_seeds(n.clone(), &neighbors, &mut points, &processed, &mut seeds);
+                while !seeds.is_empty() {
+                    seeds.sort_unstable_by(|a, b| b.cmp(a));
+                    let (i, min_point) = seeds
```

Comment:
if the seeds are sorted why query for the minimal value and not just take the first item?

Reply:
ah the above sorts indices and this the actual values